Rust/His: Speed up generalized pattern checks

beaubelgrave opened the pull request:
The existing generalized pattern check code simply linearly scans all definitions for matching packed words. This works well with a few definitions, but as more are added the performance suffers.

Create 32 lanes for UTF8 and UTF16 that split up the packed words. Use the accumulator value to find the right lane to check, then linear scan those. This makes the linear scan faster as UTF8 and UTF16 are isolated to "like" patterns.

michaelcfanning commented:
We need unit test for this change. Are we only going to support hex expressions for this argument?

beaubelgrave committed:
Rust/His: Speed up generalized pattern checks

The existing generalized pattern check code simply linearly scans all
definitions for matching packed words. This works well with a few
definitions, but as more are added the performance suffers.

Create 32 lanes for UTF8 and UTF16 that split up the packed words. Use
the accumulator value to find the right lane to check, then linear scan
those. This makes the linear scan faster as UTF8 and UTF16 are isolated
to "like" patterns.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
Rust/His: Add interop that allows specifying definitions

We don't always need to scan for everything. We often might only need to
scan for certain rules.

Update the interop in Rust to accomodate this by allowing a filter
string to be passed with a semi-colon delimited list of IDs to use.

Update the interop in C# to utilize this via the RegexPatterns that are
passed. Only the patterns that are passed are enabled.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -123,7 +125,19 @@ public void Start()
-            this.scan = identifiable_scan_create();
+            var builder = new StringBuilder();
+
+            foreach (var id in this.idToLengthMap.Keys)
```
>var [](http://example.com/codeflow?start=21&length=3)

we'd typically be explicit about type here,

beaubelgrave replied:
Moved to ThreadStatic as suggested offline.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -123,7 +125,19 @@ public void Start()
+            foreach (var id in this.idToLengthMap.Keys)
+            {
+                builder.Append(id);
+		builder.Append(';');
```
>builder [](http://example.com/codeflow?start=2&length=7)

NIT, indentation is a little off.

beaubelgrave replied:
Fixed.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -123,7 +125,19 @@ public void Start()
+            foreach (var id in this.idToLengthMap.Keys)
+            {
+                builder.Append(id);
+		builder.Append(';');
```
>builder [](http://example.com/codeflow?start=2&length=7)

nit, our convention for dropping lines on parameters is for the first arg to remain on the same line as the call, and args to left-align underneath that location.

an old Windows C++ convention i came to like. :)

beaubelgrave replied:
Fixed to just a single line, since only 2 args.

michaelcfanning commented on src/Microsoft.Security.Utilities.Core/IdentifiableScan.cs:
```diff
@@ -123,7 +128,18 @@ public void Start()
         if (this.scan == IntPtr.Zero)
         {
-            this.scan = identifiable_scan_create();
+            stringBuilder ??= new StringBuilder();
```
>StringBuilder [](http://example.com/codeflow?start=34&length=13)

this is exactly what i asked for so why would i complain? but on second reflection, probably we could have just had a per-instance item here. 

but let's leave it, it's not worth going back. thanks for acting on the notes.

michaelcfanning reviewed (approved):
:shipit:
